io/mpi/src/TMPIFile.cxx

```diff
@@ -0,0 +1,592 @@
+///
+/// \param[split] is the number of collectors to use
+
+TMPIFile::TMPIFile(const char *name, Option_t *option, Int_t split, const char *ftitle, Int_t compress)
+   : TMemFile(name, option, ftitle, compress), fSplitLevel(split), fMPIColor(0), fMPIRequest(0), fSendBuf(0)
+{
+   // check that split is set to reasonable value
+   CheckSplitLevel();
+   // split the MPI COMM_WORLD into sub-communicators
+   // one for each split level
+   SplitMPIComm();
+}
+
+////////////////////////////////////////////////////////////////////////////////
+/// TMPIFile destructor
+///
+/// This ensures the TMPIFile::Close function is called and any MPI
+/// communicators are freed.
+
+TMPIFile::~TMPIFile()
+{
+   // close the file
+   if (IsOpen())
+      Close();
+   // Close will call finalize if it has not already been called
+   // If MPI_Finalize has already been called no need for this
+   // but if the user wants to continue using MPI
+   // the sub communicators should be freed for other things
+   Int_t finalized = 0;
+   MPI_Finalized(&finalized);
+   if (not finalized and fSplitLevel > 1) {
+      MPI_Comm_free(&fSubComm);
+   }
+}
+
+////////////////////////////////////////////////////////////////////////////////
+/// As worker ranks exit, they send the collector empty messages.
+/// This counter keeps track of the number of empty messages the collector
+/// has received. Thereby the collector knows when all workers have exited
+/// and it can exit
+
+void TMPIFile::UpdateEndProcess()
+{
+   fEndProcess++;
+}
+
+////////////////////////////////////////////////////////////////////////////////
+/// This is the core of the Collector rank which listens for incoming
+/// messages from Worker ranks. The Collector
+
+void TMPIFile::RunCollector(Bool_t cache)
+{
+   // update the user set filename with the current process ID and Rank ID
+   // this ensures collectors do not overwrite one anothers files
+   this->SetOutputName();
+   Info("RunCollector", "writing to filename: %s", fMPIFilename.Data());
+   THashTable mergers;
+
+   Int_t client_Id = 0;
+
+   // loop until all other ranks in the subcommunicator have exited
+   while (fEndProcess != fMPILocalSize - 1) {
+      // Info("RunCollector","process counter %i",fEndProcess);
+      // check if message has been received
+      MPI_Status status;
+      // this call blocks until a message is received
+      MPI_Probe(MPI_ANY_SOURCE, MPI_ANY_TAG, fSubComm, &status);
+
+      // get bytes received
+      Int_t count;
+      MPI_Get_count(&status, MPI_CHAR, &count);
+      Int_t number_bytes = sizeof(char) * count;
+
+      // create buffer to receive message
+      char *buf = new char[number_bytes];
+      Int_t source = status.MPI_SOURCE;
+      Int_t tag = status.MPI_TAG;
+
+      // empty message signifies a Worker exited
+      if (number_bytes == 0) {
+         this->UpdateEndProcess();
+         MPI_Recv(buf, number_bytes, MPI_CHAR, source, tag, fSubComm, MPI_STATUS_IGNORE);
```

Comment:
I just copy-paste my reply to another change request since they are highly related:
This empty MPI message signals the end of the worker. It is necessary for the sense that the collector cannot close before all workers reach their ends. So we have a counter on the collector side that counts how many workers finished their work. Thus every time the collector "gets" an empty message, it will firstly update the counter and then receive this message.

Reply:
I will close this one and leave all discussions in the other dialog